Per broker configs should not trigger rolling upgrades

adamantal opened the pull request:
| Q               | A
| --------------- | ---
| Bug fix?        | no
| New feature?    | yes
| API breaks?     | no
| Deprecations?   | no
| Related tickets | -
| License         | Apache 2.0

### What's in this PR?
<!-- Explain the contents of the PR. Give an overview about the implementation, which decisions were made and why. -->
- Introduced `PerBrokerConfigurationState` in the `BrokerStatus` field
- If ConfigMap is updated and only per-broker configs are changed, rolling upgrade is not triggered

### Why?
<!-- Which problem does the PR fix? (Please remove this section if you linked an issue above) -->
Per broker configs can be updated through an API call which is faster and does less interruption in the cluster than restarting a broker.

### Additional context
<!-- Additional information we should know about (eg. edge cases, steps you followed to test the implementation) (Please remove this section if you don't need it) -->
The following scenarios should pass:
 - [x] Changing `Config` field in brokerConfigGroup (per-broker config) does not trigger rolling upgrade (no regression of existing behaviour)
 - [x] Changing `InternalListener` or `ExternalListener` config which does not alter `PodSpec` should not trigger rolling upgrade
 - [x] Changing SSL related configs which does not alter `PodSpec` should not trigger rolling upgrade
 - [x] In case of each valid per broker config change the `PerBrokerConfigurationState` status should change from out of sync to in sync.
 - [x] If user attempts to dynamically update to an invalid config, the status should stay in out of sync state (and the reconciler should not stuck in a loop) \*

\* Only provide validation from Kafka

### Checklist
<!-- Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields. -->

- [x] Implementation tested
- [x] Error handling code meets the [guideline](https://github.com/banzaicloud/developer-guide/blob/master/docs/coding-style/error-handling-guide.md)
- [x] Logging code meets the guideline

stoader commented on pkg/k8sutil/perbrokerconfig.go:
```diff
@@ -0,0 +1,40 @@
+package k8sutil
+
+// these configurations will not trigger rolling upgrade when updated
+var PerBrokerConfigs = []string{
```
Not sure if the `k8sutil` package is the right place for this list as it is specific to Kafka brokers. I'd place it in an internal broker config package an make it private. Since this list should have no duplicates probable is better to store it in a `map[string]struct{}`

adamantal replied:
Reconciling k8s resources are performed through `k8sutil/resources.go` 's `Reconcile` function. This function depends on the Kafka controller package, because this function sets the config in sync/out of sync states. This causes a circular dependency when I moved it into the kafka package. Where should it be put then?

stoader replied:
I'd move out the updating of the broker status when the config map changes into `kafka.go` where the config map change originates from.

adamantal replied:
Moved the related parts to the kafka package.

adamantal commented:
Updated the PR and tested it against the scenarios mentioned above.
Though the scope has shrunk since I started working on this, I think this feature provides user substantial benefits.

This patch roughly includes:
- some configurations are marked as dynamically updatable (like `listener.security.protocol.map`, `ssl.client.auth`, `listeners`, `advertised.listeners`). Updating these would not trigger rolling upgrade by itself (but if the spec of the broker pods have been changed, it will, which is 90% of the cases)
- these configs are still reflected in the configmap, but the pod will not be recreated unless a read-only config has been changed
- added a status to track the per-broker config status: in sync/out of sync
- added pre-validation before updating the per-broker configs: 
  - though honestly it is not very effective, in a lot of cases validation passes, but the broker will remain out of sync per-broker config state as broker will throw error upon the change
  - if the validation fails, the per-broker config status will be changed to error
- added special case for `listener.security.protocol.map`: we will trigger rolling upgrade based on the change, but not always
- moved post configmap update handling back to kafka package
- moved the code responsible for dynamically updating the broker config to a separate go file
- added some unit tests

Please feel free to provide your opinion the patch. This is a medium sized change, so let's be as cautious as possible.

stoader commented on pkg/k8sutil/status.go:
```diff
@@ -94,6 +94,8 @@ func generateBrokerState(brokerIds []string, cluster *banzaicloudv1beta1.KafkaCl
 				cluster.Status.BrokersState = map[string]banzaicloudv1beta1.BrokerState{brokerId: {GracefulActionState: state}}
 			case banzaicloudv1beta1.ConfigurationState:
 				cluster.Status.BrokersState = map[string]banzaicloudv1beta1.BrokerState{brokerId: {ConfigurationState: s}}
+			case banzaicloudv1beta1.PerBrokerConfigurationState:
```
I believe this code block
```
for _, brokerId := range brokerIds {
    if cluster.Status.BrokersState == nil {
        ...
    } else if val, ok := cluster.Status.BrokersState[brokerId]; ok {
        ...
    } else {
        ...
    }
}
```

can be simplified as follows:
```
brokersState := cluster.Status.BrokersState
if brokersState == nil {
    brokersState = make(map[string]banzaicloudv1beta1.BrokerState, len(brokerIds))
}

for _, brokerId := range brokerIds {
    brokerState := brokersState[brokerId]
    switch s := state.(type) {
			case banzaicloudv1beta1.RackAwarenessState:
				brokerState[brokerId] = banzaicloudv1beta1.BrokerState{RackAwarenessState: s}
			case banzaicloudv1beta1.GracefulActionState:
				brokerState[brokerId] = banzaicloudv1beta1.BrokerState{GracefulActionState: s}
			case map[string]banzaicloudv1beta1.GracefulActionState:
				state := s[brokerId]
				brokerState[brokerId] = banzaicloudv1beta1.BrokerState{GracefulActionState: state}
			case banzaicloudv1beta1.ConfigurationState:
				brokerState[brokerId] = banzaicloudv1beta1.BrokerState{ConfigurationState: s}
			case banzaicloudv1beta1.PerBrokerConfigurationState:
				brokerState[brokerId] = banzaicloudv1beta1.BrokerState{PerBrokerConfigurationState: s}
			case map[string]banzaicloudv1beta1.VolumeState:
				gracefulActionState := banzaicloudv1beta1.GracefulActionState{
					VolumeStates: s,
				}
				brokerState[brokerId] = banzaicloudv1beta1.BrokerState{GracefulActionState: gracefulActionState}
			case map[string]map[string]banzaicloudv1beta1.VolumeState:
				state := s[brokerId]
				gracefulActionState := banzaicloudv1beta1.GracefulActionState{
					VolumeStates: state,
				}
				brokerState[brokerId] = banzaicloudv1beta1.BrokerState{GracefulActionState: gracefulActionState}
            default: errors.New("unsupported broker state type")
    }
    brokersState[brokerId] = brokerState
}

cluster.Status.BrokersState = brokersState
```

stoader commented on pkg/resources/kafka/configmap.go:
```diff
@@ -250,3 +255,116 @@ func (r Reconciler) generateBrokerConfig(id int32, brokerConfig *v1beta1.BrokerC
 	return strings.Join(completeConfig, "\n")
 }
+
+func (r *Reconciler) postConfigMapReconcile(log logr.Logger, desired *corev1.ConfigMap) error {
```
This method updates broker status so updateBrokerStatus might be a more descriptive name

adamantal replied:
Renamed

stoader commented on pkg/resources/kafka/configs.go:
```diff
@@ -0,0 +1,205 @@
+	"github.com/banzaicloud/kafka-operator/pkg/util"
+)
+
+const securityProtocolMap = "listener.security.protocol.map"
```
```suggestion
const securityProtocolMapConfig = "listener.security.protocol.map"
```

stoader commented on pkg/resources/kafka/configs.go:
```diff
@@ -0,0 +1,205 @@
+	return nil
+}
+
+func getBrokerConfigsFromConfigMap(configMap *corev1.ConfigMap) map[string]string {
```
This function seems to perform the same thing as `util.ParsePropertiesFormat`.

adamantal replied:
Indeed, did not see this function. Remove the function and the corresponding tests.

stoader commented on pkg/resources/kafka/configmap.go:
```diff
@@ -250,3 +255,116 @@ func (r Reconciler) generateBrokerConfig(id int32, brokerConfig *v1beta1.BrokerC
+			if currentConfigs[securityProtocolMap] == "" || desiredConfigs[securityProtocolMap] == "" {
+				continue
+			}
+			currentListenerProtocolMap := make(map[string]string, 0)
```
```suggestion
			currentListenerProtocolMap := make(map[string]string)
```

stoader commented on pkg/resources/kafka/configmap.go:
```diff
@@ -250,3 +255,116 @@ func (r Reconciler) generateBrokerConfig(id int32, brokerConfig *v1beta1.BrokerC
+				continue
+			}
+			currentListenerProtocolMap := make(map[string]string, 0)
+			desiredListenerProtocolMap := make(map[string]string, 0)
```
```suggestion
			desiredListenerProtocolMap := make(map[string]string)
```

stoader commented on pkg/resources/kafka/configmap.go:
```diff
@@ -250,3 +255,116 @@ func (r Reconciler) generateBrokerConfig(id int32, brokerConfig *v1beta1.BrokerC
+			currentListenerProtocolMap := make(map[string]string, 0)
+			desiredListenerProtocolMap := make(map[string]string, 0)
+			for _, listenerConfig := range strings.Split(currentConfigs[securityProtocolMap], ",") {
+				listenerKeyValue := strings.Split(listenerConfig, ":")
```
I'd rename `securityProtocolMap` to `securityProtocolMapConfig` and `listenerKeyValue` to `listenerProtocolMap" to make clearer what does variables are used for.

stoader commented on pkg/resources/kafka/configmap.go:
```diff
@@ -250,3 +255,116 @@ func (r Reconciler) generateBrokerConfig(id int32, brokerConfig *v1beta1.BrokerC
+				if len(listenerKeyValue) != 2 {
+					continue
+				}
+				currentListenerProtocolMap[listenerKeyValue[0]] = listenerKeyValue[1]
```
Trim spaces on key and value.

stoader commented on pkg/resources/kafka/configmap.go:
```diff
@@ -250,3 +255,116 @@ func (r Reconciler) generateBrokerConfig(id int32, brokerConfig *v1beta1.BrokerC
+				desiredListenerProtocolMap[listenerKeyValue[0]] = listenerKeyValue[1]
+			}
+
+			for protocolName, desiredListenerProtocol := range desiredListenerProtocolMap {
```
There is no need for the `for` loop. The same verification can be done in the `for` loop above instead of populating `desiredListenerProtocolMap`

```
```

stoader commented on pkg/resources/kafka/configmap.go:
```diff
@@ -250,3 +255,116 @@ func (r Reconciler) generateBrokerConfig(id int32, brokerConfig *v1beta1.BrokerC
+	touchedConfigs := collectTouchedConfigs(currentConfigs, desiredConfigs, log)
+
+OUTERLOOP:
+	for _, config := range touchedConfigs {
```
This part could be refactored as follows to make it more readable:
```
touchedConfigs := collectTouchedConfigs(currentConfigs, desiredConfigs, log)

for _, perBrokerConfig := range perBrokerConfigs {
  for _, config := range touchedConfigs {
           if config != perBrokerConfig {
              return false // broker config that can not be updated dynamicaaly has been modified
           }

 		   if listenersSecurityProtocolChanged(...) {
                return false
           }
  }
}

return true
```

adamantal replied:
Actually if we return with a set data structure from `collectTouchedConfigs` we can do this much simpler. I come up with the following solution:
```go
	touchedConfigs := collectTouchedConfigs(currentConfigs, desiredConfigs, log)
	if _, ok := touchedConfigs[securityProtocolMapConfigName]; ok {
		if listenersSecurityProtocolChanged(currentConfigs, desiredConfigs) {
			return false
		}
	}
	for _, perBrokerConfig := range perBrokerConfigs {
		delete(touchedConfigs, perBrokerConfig)
	}
	return len(touchedConfigs) == 0
```
Assuming `collectTouchedConfigs` returns with `map[string]struct{}`

baluchicken commented on pkg/resources/kafka/configmap.go:
```diff
@@ -275,3 +280,113 @@ func (r Reconciler) generateBrokerConfig(id int32, brokerConfig *v1beta1.BrokerC
 	return strings.Join(completeConfig, "\n")
 }
+
+func (r *Reconciler) updateBrokerStatus(log logr.Logger, desired *corev1.ConfigMap) error {
```
I think this should be inside the k8sreconciler method. We already identified the broker related configmaps during an update, all these things can be simplified with ~4 lines of code if we place this inside that method. Moreover, it removes the k8s client usage from a resource related file. If we check all other resource related file their scope is to create an object only, without any k8s interaction.

adamantal replied:
Agreed. Resolve this by putting that code back, and moving config related functions to `common.go` in `kafka/util` package to prevent import cycle being created.

baluchicken reviewed (commented):
Huge thanks for the tests.
